Add a utility function to filter out elements that are not of a certain type

naoyam opened the pull request:
This adds a utility function that I think can reduce repetition. It's not yet completed, but I'd like to have feedback before going further.

We often have code like this:

```
std::vector<Val*> values;
for (auto x: values) {
  // do something only for TensorView vals
  if (x->getValType() == ValType::TensorView) {
   TensorView* x_tv = x->as<TensorView>();
   ...
  }
}
```

This PR adds `ir_utils::filterVals`, which can be used to simplify code patterns like the above:

```
for (auto x: filterVals<TensorView>(values)) {
  // x is TensorView*
 ... 
}
```

For this, I added `static constexpr ValType type` to the `Val` classes (not yet completed; only added to `IterDomain` and `TensorView` for now).

Let me know if you have any feedback. Thanks!

Commit:
Add a utility function to filter out elements that are not of a certain
type

Commit:
Fix clang-tidy

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,30 @@
+
+template <
+    typename FilterType,
+    template <typename, typename...> typename Container,
```
idea: instead of trying to fit the container type, how about we pass it more generically and use std::begin(c) ... std::end(c) to iterate it? (that's what range based for does anyway) this way it would work even with plain C arrays.

https://en.cppreference.com/w/cpp/iterator/begin

naoyam replied:
I'd love to know what that would look like. Can you show some sketch?

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,30 @@
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
+  Container<FilterType*> out;
+  for (auto& s : container) {
+    if (s->getValType() == FilterType::type) {
```
since you're interested in a better notation, would you be interested in adding another building block to the IR root class? ie. add this to Statement:

```cpp
 template<class T>
 bool IsA() const {
 return dynamic_cast<const T*>(this) != nullptr;
 }
```

this is a counterpart to Statement::as<T> that I have on my todo list, and would allow us to say:

`if (s->isA<FilterType>()) { ... }`

... and also eventually eliminate the need for ValType altogether.

naoyam replied:
That looks nice, but one thing I didn't want to change is that the type checking is currently always done with `getValType()`. I think we can replace that with something like `IsA`, but I wonder whether `ValType` values and classes have one-to-one relationship. For example, would there be a case where there could be two different classes that are a subclass of `Val` AND have the same `ValType` value. I don't think that's the case now, but is this something that could happen in the future?

jjsjann123 replied:
I think switch to a more consistent `IsA` makes sense to me (uniformed API).
Can't we have specialization of `isA` so that we cannot necessarily do dynamic cast? that seems to also resolve the restriction on the one-to-one mapping.

```
 template<class T>
 bool IsA() const {
 return dynamic_cast<const T*>(this) != nullptr;
 }

 template<ValType T>
 bool IsA() const {
 return getValType() == T;
 }
```

jjsjann123 replied:
Using `IsA<>()` with the non-mapping filter and return would still require us to add an extra template parameter in the filterVals, (one for filter criterion used in the `IsA`, one to cast the return type `as`). Might bother people :shrug:

tlemo replied:
@naoyam: a potential mismatch between the tag (`ValType`) and the actual type is one of the reasons I'm wary of using tags  , especially when they duplicate the type hierarchy 1:1 (like we do today). Since we're using a hierarchy to model the taxonomy of nodes, I think it's both easier and safer to use the types themselves. Another benefit of using the types (through a notation like `isA<T>`), is that you can query for abstract (non-leaf) types too, ex. `isA<Expr>` or `isA<Val>`

@jjsjann123 : as I explained above, the main benefit of `isA<T>` is not the notation itself, but the fact that it avoids tags altogether. So a version which tries to mix tags and types seems confusing at best and I'd recommend against it.

I'm not suggesting we should go as far as eliminating tags in this change. It seems an opportunity for adding `isA<T>` since it fits well with the proposed utility, but it's not really critical to this PR.

naoyam replied:
I agree that we could get rid of the tag and just use the runtime type information would be more robust. However, I'd prefer to keep using the tag for type checking at least in this PR as changing that is not the purpose. I'll create a separate issue to consider.

naoyam replied:
See #268

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,30 @@
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
+  Container<FilterType*> out;
+  for (auto& s : container) {
```
nit: using the explicit root type here woudl avoid the need for the `template as<FilterType>` below

naoyam replied:
Sorry, can you give me more details?

tlemo replied:
I just meant `for (Statement* s : container) ...`.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,30 @@
+    typename FilterType,
+    template <typename, typename...> typename Container,
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
```
why does it have to return the same container as the source?

naoyam replied:
Just for simplicity.

tlemo replied:
Wouldn't be simpler (for the non-iterator version) to just return a `std::vector<FilterType*>`? That avoids the whole complication with trying to deduce the container type, and leads to more predictable behavior too (we wouldn't necessarily want to use a std::list or std::deque to return the results just to match the original container)

Commit:
Custom container-based version

Commit:
clang-format

jjsjann123 commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+class FilterValContainer {
+ public:
+  using value_type = FilterType*;
+  using const_iterator = FilterIterator<FilterType, InputIt>;
```
nitpick: Wouldn't we also want a non-const iterator?

tlemo replied:
Do we want to modify the original container through the filtered view?

naoyam replied:
There may be some use cases, but I suspect we don't need that at this time. I left a related comment for a future extension.

jjsjann123 reviewed (commented):
I like the updated version a lot :+1:

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+};
+
+template <typename FilterType, typename InputIt>
+class FilterValContainer {
```
We should probably rename this to say "view" to better reflect what it is (it's not really a container since it doesn't own the values). `FilteredView` ?

Also, `InputIt` -> just `Iterator`? (technically it's not an input iterator but rather a forward iterator)

naoyam replied:
Thanks for the suggestion. Renamed to `FilteredView`.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+  }
+
+ private:
+  InputIt input_it_;
```
1. `const` for the data members?

2. Use more precise names for the iterators: `begin_` and `end_` ? (in particular, last != end so we should avoid using that name)

naoyam replied:
Thanks. Done.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+
+template <typename FilterType, typename ContainerType>
+auto filterVals(const ContainerType& inputs) {
+  return filterVals<FilterType>(inputs.begin(), inputs.end());
```
nit: `cbegin`, `cend`

naoyam replied:
Changed.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+  return out;
+}
+#else
+template <typename FilterType, typename InputIt>
```
`InputIt` -> `IteratorType` or just `Iterator` ?

naoyam replied:
Renamed to `Iterator`.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+  }
+
+ private:
+  InputIt input_it_;
```
please separate the data members from private functions into separate `private` sections

naoyam replied:
Done.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+    return (*input_it_)->template as<FilterType>();
+  }
+
+  FilterIterator& operator++() {
```
we need a post increment implementation too

naoyam replied:
Done

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+
+ private:
+  InputIt input_it_;
+  InputIt last_;
```
1. rename input_it_ -> current_ ?
2. as I mentioned in another comment, we should rename last_ -> end_

naoyam replied:
Done.

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  bool operator==(const FilterIterator& other) const {
+    return input_it_ == other.input_it_ && last_ == other.last_;
```
I'd actually just `TORCH_INTERNAL_ASSERT(last_ == other.last_)` - if someone is trying to compare fitered iterators coming from different filtered views it's most likely an error and we should raise it as such

naoyam replied:
That's a good idea. Done.

Commit:
Remove deprecated code

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,101 @@
+
+  FilterIterator& operator++() {
+    ++input_it_;
+    advance();
```
could we use `std::find_if` instead?

naoyam replied:
Thanks for the suggestion. Simplified the implementation of `advance` with `find_if`.

Commit:
review feedback

Commit:
Add a comment on non-const iterator

Commit:
Use find_if

Commit:
Add more overloads and switch to use dynamic_cast instead of the tag

We eventually do casting so dynamic_cast-based type checking seems more
appropriate

Commit:
Use filterVals

Commit:
Remove static type field

naoyam commented:
I actually changed the type checking to use `dynamic_cast` instead of `getValType()` as it eventually does `dynamic_cast`. I think it makes more sense.

jjsjann123 reviewed (approved):
LGTM

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,107 @@
+    return (*current_)->template as<FilterType>();
+  }
+
+  FilterType* operator->() const {
```
nice, good catch!

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  FilterIterator operator++(int) {
+    auto before_increment = *this;
```
nit: `const` ?

naoyam replied:
Done

tlemo commented on torch/csrc/jit/codegen/cuda/ir_utils.h:
```diff
@@ -0,0 +1,107 @@
+    return !(*this == other);
+  }
+
+ private:
```
please place the private state section last (public methods, private methods, private state)

naoyam replied:
Done

Commit:
cleanup based on review feedback

Commit:
WIP: unit test

naoyam commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -781,6 +782,50 @@ void testGPU_FusionTensor() {
+#if 0
+  std::vector<TensorView*> tvs(
+      ir_utils::filterVals<TensorView>(vals).begin(),
+      ir_utils::filterVals<TensorView>(vals).end());
```
@tlemo Do you know why this doesn't work?

jjsjann123 replied:
Look at constructor (5) in https://en.cppreference.com/w/cpp/container/vector/vector
Your const_iterator needs to satisfy `LegacyInputIterator`... So you would need `difference_type` and bunch of that...

Commit:
Expand the unit test

Commit:
Make FilterIterator compatible as std::forward_iterator

tlemo commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -781,6 +782,40 @@ void testGPU_FusionTensor() {
+  TORCH_CHECK(ints.size() == 2);
+  TORCH_CHECK(ints[0] == scalar1);
+  TORCH_CHECK(ints[1] == scalar2);
+  return;
```
nit: `return` not needed

tlemo commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -781,6 +782,40 @@ void testGPU_FusionTensor() {
+  auto scalar1 = new Int(0);
+  auto scalar2 = new Int(1);
+
+  std::vector<Val*> vals = {tv0, scalar0, tv1, scalar1, scalar2};
```
can you please make vals `const`?

tlemo commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -781,6 +782,40 @@ void testGPU_FusionTensor() {
+  TORCH_CHECK(floats.size() == 1);
+  TORCH_CHECK(floats[0] == scalar0);
+
+  std::vector<Int*> ints(
```
one more test case please: a filtering which returns an empty list, ex:

```
for (auto ptr : ir_utils::filterVals<Expr>(vals)) {
 TORCH_CHECK(false, "Not expecting any results");
}

```

Commit:
Rename filterVals to filterByType as it is no longer specific to Val

naoyam commented:
I renamed the filter function to `filterByType` as it is actually no longer specific to `Val`. It just checks the runtime type of each element and filters out anything that is of that type.

The example I gave first still looks mostly the same. We replace this:

```
std::vector<Val*> values;
for (Val* x: values) {
  // do something only for TensorView vals
  if (x->getValType() == ValType::TensorView) {
   TensorView* x_tv = x->as<TensorView>();
   ...
  }
}
```

with:

```
for (TensorView* x: filterByType<TensorView>(values)) {
  // x is TensorView*
 ... 
}
```

naoyam commented:
@tlemo Sorry, I missed your last comments. I'll update the test with a separate PR.
